keps/sig-node/20191118-cgroups-v2.md

```diff
@@ -0,0 +1,216 @@
+
+### Current status of dependencies
+
+- CRI-O+crun: support cgroups v2
+
+- runc support for cgroups v2 is work in progress
+
+- containerd: [https://github.com/containerd/containerd/pull/3799](https://github.com/containerd/containerd/pull/3799)
+
+- Moby: [https://github.com/moby/moby/pull/40174](https://github.com/moby/moby/pull/40174)
+
+- OCI runtime spec: TODO
+
+## Current cgroups usage and the equivalent in cgroups v2
+
+|Kubernetes cgroups v1|Kubernetes cgroups v2 behavior|
+|---|---|
+|CPU stats for Horizontal Pod Autoscaling|No .percpu cpuacct stats.  Check cadvisor.|
+|CPU pinning based on integral cores|Cpuset controller available|
+|Memory limits|Not changed, different naming|
+|PIDs limits|Not changed, same naming|
+|hugetlb|Not present|
+
+## Proposal
+
+The proposal is to implement cgroups v2 in two different ways.
+
+The first one ensures that any configuration file designed for cgroups
+v1 will continue to work on cgroups v2.
+
+The second one requires changes through the entire stack, including
+the OCI runtime specifications.
+
+At startup the Kubelet detects what hierarchy the system is using.  It
+checks the file system type for `/sys/fs/cgroup` (the equivalent of
+`stat -f --format '%T' /sys/fs/cgroup`).  If the type is `cgroup2fs`
+then the Kubelet will use only cgroups v2 during all its execution.
+
+### Option 1: Convert from cgroups v1 settings to v2
+
+We can convert the values passed by the k8s in cgroups v1 from to
+cgroups v2 so Kubernetes users don’t have to change what they specify
+in their manifests.
+
+crun has implemented the conversion as follows:
+
+**Memory controller**
+
+| OCI (x) | cgroup 2 value (y) | conversion  |   comment |
+|---|---|---|---|
+| limit | memory.max | y = x ||
+| swap | memory.swap_max | y = x ||
+| reservation | memory.low | y = x ||
+
+**PIDs controller**
+
+| OCI (x) | cgroup 2 value (y) | conversion  |   comment |
+|---|---|---|---|
+| limit | pids.max | y = x ||
+
+**CPU controller**
+
+| OCI (x) | cgroup 2 value (y) | conversion  |  comment |
+|---|---|---|---|
+| shares | cpu.weight | y = (1 + ((x - 2) * 9999) / 262142) | convert from [2-262144] to [1-10000]|
+| period | cpu.max | y = x| period and quota are written together|
+| quota | cpu.max | y = x| period and quota are written together|
+
+**blkio controller**
+
+| OCI (x) | cgroup 2 value (y) | conversion  |   comment |
+|---|---|---|---|
+| weight | io.bfq.weight | y = (1 + (x - 10) * 9999 / 990) | convert linearly from [10-1000] to [1-10000]|
+| weight_device | io.bfq.weight | y = (1 + (x - 10) * 9999 / 990) | convert linearly from [10-1000] to [1-10000]|
+|rbps|io.max|y=x||
+|wbps|io.max|y=x||
+|riops|io.max|y=x||
+|wiops|io.max|y=x||
+
+**cpuset controller**
+
+| OCI (x) | cgroup 2 value (y) | conversion  |   comment |
+|---|---|---|---|
+| cpus | cpuset.cpus | y = x ||
+| mems | cpuset.mems | y = x ||
+
+
+With this approach cAdvisor would have to read back values from
+cgroups v2 files (already done).
```

Comment:
So, with this option, CRI implementations will need to map the cgroup v1 fields to cgroup v2 values, right? 

Could you list what each layer needs to do for each option to make it clear?
 - user (pod spec)
 - kubelet
 - CRI implementation
 - OCI runtime

Reply:
the changes in the CRI implementation and OCI runtime might be implemented in a different way, deciding where to draw the line.
For CRI-O+crun, most of the logic is in the OCI runtime itself, but that is not the only way to achieve it.
For the Kubelet, the changes in this patch should be enough (at least until we have not hugetlb available): https://github.com/kubernetes/kubernetes/pull/85218

In the second phase though, when cgroup v2 is fully supported through the stack there is need to change both pod specs+CRI.

Reply:
can you clarify why you feel the pod spec needs to change?  i see no major reason to change the pod spec or resource representation.

Reply:
I was assuming we'd like to expose all/most of the cgroup v2 features.

e.g. the memory controller on cgroup v1 allows to configure:
- `memory.soft_limit_in_bytes`
- `memory.limit_in_bytes`

while on cgroup v2 we have:
- `memory.high`
- `memory.low`
- `memory.max`
- `memory.min`

but that is probably out of the scope as each new feature (if needed) must go through its own KEP?

Reply:
@derekwaynecarr are future improvements based on what cgroup 2 offers out of scope for the current KEP?

Reply:
@giuseppe future improvements for what cgroup v2 offers should be out of scope of this kep.  i would keep this kep focused on kubelet is tolerant of cgroup v2 host.  adding new cgroup v2 specific features to resource model would be a separate enhancement.

To address @yujuhong question, I think we are saying the following:

- no change required to pod spec
- kubelet cgroup manager uses v1 or v2 mode based on its introspection of sys/fs on startup
- kubelet qos and pod level cgroup creation on a v2 host uses the mapping table specified
- cri implementers that support cgroup v2 hosts must use a similar mapping to ensure that pod bounding cgroup values are consistent with the container cgroup values.
- oci runtime spec is not required to change as cri implementer provides mapping in the transitional period.

@giuseppe agree with above?